singleflight/example_test.go

```diff
@@ -0,0 +1,47 @@
+/*
+Copyright 2016 The Go Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package singleflight_test
+
+import (
+	"fmt"
+	"log"
+	"sync/atomic"
+	"time"
+
+	"github.com/golang/groupcache/singleflight"
+)
+
+var counter int32
+var networkGroup singleflight.Group
+
+// expensiveNetworkCall simulates an expensive action.
+func expensiveNetworkCall() (interface{}, error) {
+	time.Sleep(1)
+	return atomic.AddInt32(&counter, 1), nil
+}
+
+func Example() {
+	// No matter how many threads call Do, only one network call per key will
+	// be in progress at any time. Callers that share a key will get the same
+	// results as an in-flight call with that key.
+	v, err := networkGroup.Do("key", expensiveNetworkCall)
```

Comment:
Normally this expensiveNetworkCall is a closure, closing over the `key` value. Maybe show an addition of +1 instead and say:

```go
    // Imagine this is something expensive:
    return n + 1, nil
```

And make your "key" be:

    n := 41
    group.Do(strconv.Itoa(n), func(...) {..

(no need to mention 'network')